Incorrect $TMPDIR environment variable on OS X

## Description

We're using chef homebrew_package resource to compile and install package: 
- chef executes `brew install <formula>` command from user, defined as homebrew_user
- brew downloads sources and tries to run configure/make, for temporary files it uses directory defined in $TMPDIR environment variable (or /tmp is $TMPDIR not exist), but in this case $TMPDIR points to temporary files directory of another user (vagrant)
```
bash-3.2# ls -la /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T/
total 0
drwx------  2 vagrant  staff   68 Mar 15 22:11 .
drwxr-xr-x  5 vagrant  staff  170 Jan 25 10:57 ..
```

OS X DARWIN_USER_TEMP_DIR (which is set to TMPDIR) is designed to use separate random generated directories for each user, so when we're changing user (or elevating privileges) we should unset it or change to value returned by `getconf DARWIN_USER_TEMP_DIR` (detailed info in References)

In this case we have bug at [vagrant side](https://github.com/mitchellh/vagrant/issues/7140) - vagrant should unset this variable when elevating privileges from vagrant user to root, and at chef side too - chef should unset $TMPDIR value or change it to `getconf DARWIN_USER_TEMP_DIR` output, when running brew install command from regular user

## Chef Version

12.6.0, 12.8.1

## Platform Version

OS X 10.11.2 (15C50) with chef 12.6.0 and 12.8.1

## Replication Case

using vagrant box (on OS X 10.11.3 (15D21)) with following vagrant provisioner configuration:
```ruby
  config.vm.provision 'shell', inline: <<-SHELL
    env | grep TMPDIR
    echo "homebrew_package 'wget' do
        homebrew_user 'brew_user'
        options '--build-from-source'
    end" > /tmp/pack.rb
    chef-apply /tmp/pack.rb
    SHELL
```

## Client Output

```
==> default: Running provisioner: shell...
    default: Running: inline script
==> default: TMPDIR=/var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T/
==> default: [2016-03-15T22:36:02+00:00] INFO: Run List is []
==> default: [2016-03-15T22:36:02+00:00] INFO: Run List expands to []
==> default: [2016-03-15T22:36:02+00:00] INFO: Processing homebrew_package[wget] action install ((chef-apply cookbook)::(chef-apply recipe) line 1)
==> default:
==> default: ================================================================================
==> default: Error executing action `install` on resource 'homebrew_package[wget]'
==> default: ================================================================================
==> default:
==> default: Mixlib::ShellOut::ShellCommandFailed
==> default: ------------------------------------
==> default: Expected process to exit with [0], but received '1'
==> default: ---- Begin output of brew install --build-from-source wget ----
==> default: STDOUT: ==> Installing dependencies for wget: xz, pkg-config, makedepend
==> default: ==> Installing wget dependency: xz
==> default: ==> Downloading https://fossies.org/linux/misc/xz-5.2.2.tar.gz
==> default: ==> ./configure --disable-silent-rules --prefix=/usr/local/Cellar/xz/5.2.2
==> default: Last 15 lines from /Users/brew_user/Library/Logs/Homebrew/xz/01.configure:
==> default:
==> default: ./configure
==> default: --disable-debug
==> default: --disable-dependency-tracking
==> default: --disable-silent-rules
==> default: --prefix=/usr/local/Cellar/xz/5.2.2
==> default:
==> default:
==> default: XZ Utils 5.2.2
==> default:
==> default: System type:
==> default: checking build system type... mkdir: /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T//cg29037-1360: Permission denied
==> default: mkdir: /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T//cg-29037: Permission denied
==> default: config.guess: cannot create a temporary directory in /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T/
==> default: configure: error: cannot guess build type; you must specify one
==> default:
==> default: READ THIS: https://git.io/brew-troubleshooting
==> default: STDERR:
==> default: ---- End output of brew install --build-from-source wget ----
==> default: Ran brew install --build-from-source wget returned 1
==> default:
==> default: Resource Declaration:
==> default: ---------------------
==> default: # In /tmp/pack.rb
==> default:
==> default:   1: homebrew_package 'wget' do
==> default:   2:         homebrew_user 'brew_user'
==> default:   3:         options '--build-from-source'
==> default:   4:     end
==> default:
==> default: Compiled Resource:
==> default: ------------------
==> default: # Declared in /tmp/pack.rb:1:in `run_chef_recipe'
==> default:
==> default: homebrew_package("wget") do
==> default:   package_name "wget"
==> default:   action [:install]
==> default:   retries 0
==> default:   retry_delay 2
==> default:   default_guard_interpreter :default
==> default:   declared_type :homebrew_package
==> default:   cookbook_name "(chef-apply cookbook)"
==> default:   recipe_name "(chef-apply recipe)"
==> default:   homebrew_user "brew_user"
==> default:   options "--build-from-source"
==> default: end
==> default:
==> default: [2016-03-15T22:36:09+00:00] INFO: Running queued delayed notifications before re-raising exception
==> default: [2016-03-15T22:36:09+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
==> default: [2016-03-15T22:36:09+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
==> default: [2016-03-15T22:36:09+00:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: homebrew_package[wget] ((chef-apply cookbook)::(chef-apply recipe) line 1) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
==> default: ---- Begin output of brew install --build-from-source wget ----
==> default: STDOUT: ==> Installing dependencies for wget: xz, pkg-config, makedepend
==> default: ==> Installing wget dependency: xz
==> default: ==> Downloading https://fossies.org/linux/misc/xz-5.2.2.tar.gz
==> default: ==> ./configure --disable-silent-rules --prefix=/usr/local/Cellar/xz/5.2.2
==> default: Last 15 lines from /Users/brew_user/Library/Logs/Homebrew/xz/01.configure:
==> default:
==> default: ./configure
==> default: --disable-debug
==> default: --disable-dependency-tracking
==> default: --disable-silent-rules
==> default: --prefix=/usr/local/Cellar/xz/5.2.2
==> default:
==> default:
==> default: XZ Utils 5.2.2
==> default:
==> default: System type:
==> default: checking build system type... mkdir: /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T//cg29037-1360: Permission denied
==> default: mkdir: /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T//cg-29037: Permission denied
==> default: config.guess: cannot create a temporary directory in /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T/
==> default: configure: error: cannot guess build type; you must specify one
==> default:
==> default: READ THIS: https://git.io/brew-troubleshooting
==> default: STDERR:
==> default: ---- End output of brew install --build-from-source wget ----
==> default: Ran brew install --build-from-source wget returned 1
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

## Stacktrace

```
Generated at 2016-03-15 22:36:09 +0000
Mixlib::ShellOut::ShellCommandFailed: homebrew_package[wget] ((chef-apply cookbook)::(chef-apply recipe) line 1) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process
 to exit with [0], but received '1'
---- Begin output of brew install --build-from-source wget ----
STDOUT: ==> Installing dependencies for wget: xz, pkg-config, makedepend
==> Installing wget dependency: xz
==> Downloading https://fossies.org/linux/misc/xz-5.2.2.tar.gz
==> ./configure --disable-silent-rules --prefix=/usr/local/Cellar/xz/5.2.2
Last 15 lines from /Users/brew_user/Library/Logs/Homebrew/xz/01.configure:

./configure
--disable-debug
--disable-dependency-tracking
--disable-silent-rules
--prefix=/usr/local/Cellar/xz/5.2.2


XZ Utils 5.2.2

System type:
checking build system type... mkdir: /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T//cg29037-1360: Permission denied
mkdir: /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T//cg-29037: Permission denied
config.guess: cannot create a temporary directory in /var/folders/f4/xjctx1391rbdftkrjkbzjp7c0000gp/T/
configure: error: cannot guess build type; you must specify one

READ THIS: https://git.io/brew-troubleshooting
STDERR:
---- End output of brew install --build-from-source wget ----
Ran brew install --build-from-source wget returned 1
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/mixlib-shellout-2.2.6/lib/mixlib/shellout.rb:289:in `invalid!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/mixlib-shellout-2.2.6/lib/mixlib/shellout.rb:276:in `error!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/mixin/shell_out.rb:56:in `shell_out!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package.rb:538:in `shell_out_with_timeout!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package/homebrew.rb:130:in `get_response_from_command'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package/homebrew.rb:76:in `brew'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package/homebrew.rb:49:in `install_package'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package.rb:100:in `block (2 levels) in action_install'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package.rb:226:in `multipackage_api_adapter'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package.rb:99:in `block in action_install'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/mixin/why_run.rb:52:in `add_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider.rb:175:in `converge_by'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/package.rb:98:in `action_install'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider.rb:144:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource.rb:596:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:73:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `each'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_list.rb:84:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_list.rb:82:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:104:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application/apply.rb:195:in `run_chef_recipe'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application/apply.rb:205:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application/apply.rb:218:in `run'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/bin/chef-apply:25:in `<top (required)>'
/usr/local/bin/chef-apply:50:in `load'
/usr/local/bin/chef-apply:50:in `<main>'
```

## Workaround

* We can use `unset TMPDIR` in provisioning vagrant script before running chef-client/chef-apply
```ruby
  config.vm.provision 'shell', inline: <<-SHELL
    env | grep TMPDIR
    unset TMPDIR
    env | grep TMPDIR
    echo "homebrew_package 'wget' do
        homebrew_user 'brew_user'
        options '--build-from-source'
    end" > /tmp/pack.rb
    chef-apply /tmp/pack.rb
    SHELL
```
* Another variant: use 'vagrant' as homebrew_user

## References
- http://www.magnusviri.com/OS_X_Admin/what-is-var-folders.html
- http://apple.stackexchange.com/a/131080

--- a/lib/chef/provider/package/homebrew.rb
+++ b/lib/chef/provider/package/homebrew.rb
@@ -127,7 +127,7 @@ class Chef
 
           Chef::Log.debug "Executing '#{command}' as user '#{homebrew_user.name}'"
           # FIXME: this 1800 second default timeout should be deprecated
-          output = shell_out_with_timeout!(command, :timeout => 1800, :user => homebrew_uid, :environment => { "HOME" => homebrew_user.dir, "RUBYOPT" => nil })
+          output = shell_out_with_timeout!(command, :timeout => 1800, :user => homebrew_uid, :environment => { "HOME" => homebrew_user.dir, "RUBYOPT" => nil, "TMPDIR" => nil })
           output.stdout.chomp
         end
 

